tests/test_config.py

```diff
@@ -0,0 +1,314 @@
+
+    # Check that load_in_config_file fails as expected
+    with pytest.raises(expected_exception):
+        config.load_in_config_files(files)
+
+    log_check(caplog, expected_log_entries)
+
+
+@pytest.mark.parametrize(
+    "config_dict,expected_exception,expected_log_entries",
+    [
+        (
+            {"core": {"grid": {"nx": 10, "ny": 10}}},
+            KeyError,
+            (
+                (
+                    CRITICAL,
+                    "Core configuration does not specify which other modules should be "
+                    "configured!",
+                ),
+            ),
+        ),
+        (
+            {"core": {"modules": ["soil", "soil"]}},
+            RuntimeError,
+            (
+                (
+                    CRITICAL,
+                    "The list of modules to configure given in the core configuration "
+                    "file repeats 1 names!",
+                ),
+            ),
+        ),
+    ],
+)
+def test_find_schema(caplog, config_dict, expected_exception, expected_log_entries):
+    """Check errors in finding module schema."""
+
+    # Check that find_schema fails as expected
+    with pytest.raises(expected_exception):
+        config.find_schema(config_dict)
+
+    log_check(caplog, expected_log_entries)
+
+
+@pytest.mark.parametrize(
+    "modules,expected_exception,expected_log_entries",
+    [
+        (
+            ["a_stupid_module_name"],
+            RuntimeError,
+            (
+                (
+                    CRITICAL,
+                    "Expected a schema for a_stupid_module_name module configuration, "
+                    "it was not provided!",
+                ),
+            ),
+        ),
+    ],
+)
+def test_construct_combined_schema(
+    caplog, modules, expected_exception, expected_log_entries
+):
+    """Checks errors for bad or missing json schema."""
+
+    # Check that construct_combined_schema fails as expected
+    with pytest.raises(expected_exception):
+        config.construct_combined_schema(modules)
+
+    log_check(caplog, expected_log_entries)
+
+
+@pytest.mark.parametrize(
+    "expected_log_entries",
+    [
+        (
+            (
+                (
+                    INFO,
+                    "Configuration files successfully validated!",
+                ),
+                (
+                    INFO,
+                    "Saving all configuration details to ./complete_config.toml",
+                ),
+            )
+        ),
+    ],
+)
+def test_final_validation_log(caplog, expected_log_entries):
+    """Checks that validation passes as expected and produces the correct output."""
+
+    config.validate_config(["tests/fixtures"], out_file_name="complete_config")
+
+    # Remove generated output file
+    # As a bonus tests that output file was generated correctly + to the right location
+    Path("./complete_config.toml").unlink()
+
+    print(caplog)
```

Comment:
Provably needs to be removed. Print statements in tests are useful only for debug purposes

Reply:
Ahh yeah I missed that!